diff -up ./bash-4.2/print_cmd.c-orig ./bash-4.2/print_cmd.c
--- ./bash-4.2/print_cmd.c-orig	2013-12-04 13:22:53.850843813 +0100
+++ ./bash-4.2/print_cmd.c	2013-12-04 13:23:28.249867347 +0100
@@ -1394,7 +1394,7 @@ indent (amount)
   for (i = 0; amount > 0; amount--)
     indentation_string[i++] = ' ';
   indentation_string[i] = '\0';
-  cprintf (indentation_string);
+  cprintf ("%s", indentation_string);
 }
 
 static void
