diff -up bash-4.0/doc/bash.1.ulimit-m bash-4.0/doc/bash.1
--- bash-4.0/doc/bash.1.ulimit-m	2009-01-21 16:31:19.000000000 +0100
+++ bash-4.0/doc/bash.1	2009-01-21 16:31:19.000000000 +0100
@@ -9044,7 +9044,7 @@ The maximum number of pending signals
 The maximum size that may be locked into memory
 .TP
 .B \-m
-The maximum resident set size
+The maximum resident set size (has no effect on Linux)
 .TP
 .B \-n
 The maximum number of open file descriptors (most systems do not
diff -up bash-4.0/builtins/ulimit.def.ulimit-m bash-4.0/builtins/ulimit.def
--- bash-4.0/builtins/ulimit.def.ulimit-m	2009-01-04 20:32:23.000000000 +0100
+++ bash-4.0/builtins/ulimit.def	2009-01-21 16:36:12.000000000 +0100
@@ -40,7 +40,7 @@ Options:
   -f	the maximum size of files written by the shell and its children
   -i	the maximum number of pending signals
   -l	the maximum size a process may lock into memory
-  -m	the maximum resident set size
+  -m	the maximum resident set size (has no effect on Linux)
   -n	the maximum number of open file descriptors
   -p	the pipe buffer size
   -q	the maximum number of bytes in POSIX message queues
