diff -up bash-4.0/assoc.c.key_alloc bash-4.0/assoc.c
--- bash-4.0/assoc.c.key_alloc	2009-08-26 16:17:50.000000000 +0200
+++ bash-4.0/assoc.c	2009-08-26 16:18:42.000000000 +0200
@@ -77,6 +77,7 @@ assoc_insert (hash, key, value)
   b = hash_search (key, hash, HASH_CREATE);
   if (b == 0)
     return -1;
+  b->key = savestring (key);
   FREE (b->data);
   b->data = value ? savestring (value) : (char *)0;
   return (0);
