diff -up bash-4.2/general.h.patch bash-4.2/general.h
--- bash-4.2/general.h.patch	2012-08-21 11:34:25.978724313 +0200
+++ bash-4.2/general.h	2012-08-21 11:34:52.235947602 +0200
@@ -104,7 +104,7 @@ extern char *strcpy __P((char *, const c
 
 /* Define exactly what a legal shell identifier consists of. */
 #define legal_variable_starter(c) (ISALPHA(c) || (c == '_'))
-#define legal_variable_char(c)	(ISALNUM(c) || c == '_')
+#define legal_variable_char(c)	(ISALNUM(c) || c == '_'|| c == '.')
 
 /* Definitions used in subst.c and by the `read' builtin for field
    splitting. */
