*** ../bash-4.2/lib/glob/gmisc.c	2011-02-05 16:11:17.000000000 -0500
--- lib/glob/gmisc.c	2011-02-15 20:50:43.000000000 -0500
***************
*** 126,129 ****
--- 126,130 ----
  		{
  	          matlen += wpat - wbrack - 1;	/* incremented below */
+ 	          wpat--;			/* back up to NUL */
  	          break;
  	        }
***************
*** 262,265 ****
--- 263,267 ----
  		{
  	          matlen += pat - brack - 1;	/* incremented below */
+ 	          pat--;			/* back up to NUL */
  	          break;
  	        }
