diff -up bash-4.0/pcomplete.c.parser_state bash-4.0/pcomplete.c
--- bash-4.0/pcomplete.c.parser_state	2009-02-25 15:33:16.000000000 +0100
+++ bash-4.0/pcomplete.c	2009-02-25 15:34:06.000000000 +0100
@@ -1032,6 +1032,7 @@ gen_shell_function_matches (cs, text, li
   cmdlist = build_arg_list (funcname, text, lwords, cw);
 
   pps = &ps;
+  save_parser_state (pps);
   begin_unwind_frame ("gen-shell-function-matches");
   add_unwind_protect (restore_parser_state, (char *)pps);
   add_unwind_protect (dispose_words, (char *)cmdlist);
